File: haskell-server/src/BoardGame/Server/Service/GameData.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}

module BoardGame.Server.Service.GameData where

import Data.Sequence (Seq)
import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)

import BoardGame.Server.Domain.Play (Play)
import BoardGame.Server.Domain.GameBase (GameBase)

-- | Storable/retrievable data about a game in its entirety.
--   Includes the initial condition (base) and the sequence of successive plays.
data GameData = GameData {
    base :: GameBase
  , plays :: Seq Play
}
  deriving (Eq, Show, Generic)

instance FromJSON GameData
instance ToJSON GameData

File: scala-server/app/com/bolour/boardgame/scala/server/service/GameData.scala
package com.bolour.boardgame.scala.server.service

import com.bolour.boardgame.scala.server.domain.{Game, GameBase, Play}

import scala.util.Try

/**
  * History of the plays in the game including its initial state.
  *
  * @param base The initial state.
  * @param plays The successive plays.
  */
case class GameData(base: GameBase, plays: Vector[Play])

object GameData {
  def fromGame(game: Game): GameData = GameData(game.gameBase, game.plays)

  // TODO. Implement Game.fromTransitions - recover the current state of the game from its history.ß´
  def toGame(gameData: GameData): Try[Game] = ???

}
